src/schema/document/value.rs

```diff
@@ -9,12 +9,11 @@ use crate::tokenizer::PreTokenizedString;
 /// A single field value.
 pub trait Value<'a>: Send + Sync + Debug {
     /// The child value type returned by this doc value.
-    type ChildValue: Value<'a>;
     /// The iterator for walking through the elements within the array.
-    type ArrayIter: Iterator<Item = ReferenceValue<'a, Self::ChildValue>>;
+    type ArrayIter: Iterator<Item = Self>;
     /// The visitor walking through the key-value pairs within
     /// the object.
-    type ObjectIter: Iterator<Item = (&'a str, ReferenceValue<'a, Self::ChildValue>)>;
+    type ObjectIter: Iterator<Item = (&'a str, Self)>;
```

Comment:
The reason why this was done is because it creates huge ergonomic issues when you are building more complex indexing structures, or you have anything touching inline values, i.e. SmallVec. I don't really agree with this change as it significantly hinders ergonomics and any optimisations you might want to do when working with custom types.

Reply:
This still compiles. The PR removes a case where your value's recursive ChildValue is not self. I don't know if that's really relevant.

```rust
#[derive(Clone, Debug, Default)]
pub struct TantivyDocument2 {
    field_values: SmallVec<[FieldValue; 4]>,
}

impl Document for TantivyDocument2 {
    type Value<'a> = &'a OwnedValue;
    type FieldsValuesIter<'a> = FieldValueIter<'a>;

    fn iter_fields_and_values(&self) -> Self::FieldsValuesIter<'_> {
        FieldValueIter(self.field_values.iter())
    }
}

impl DocumentDeserialize for TantivyDocument2 {
    fn deserialize<'de, D>(mut _deserializer: D) -> Result<Self, DeserializeError>
    where
        D: DocumentDeserializer<'de>,
    {
        unimplemented!()
    }
}
```

Reply:
The above example doesn't touch on the recursive structure of the `Value` which is why it compiles. The `ChildValue` is useful in places where you have something like a flattened document where you may want a SmallVec to store the single value fields inline, which you can't do without creating a new type to prevent the recursive type error.

Reply:
Can you give an example?

Reply:
I will try to do an example later today if I have a chance, I admit it's a somewhat niche situation.